### PR TITLE
Improved Pretty Printing

### PR DESCRIPTION
**Pull request by mlutze**
Still in early stages. I'm experimenting with a couple things, including an implicit "Readability" parameter, so we'd just declare an `implicit val readability: Readability = Readability.External` where strings for the user are built and `... Readability.Internal` where strings for compiler programmers are built, and then we'd avoid potential messiness of passing a boolean in every time. But it might add unnecessary complexity; I'm not sure.

### CODE REVIEW

**Comment by mlutze** [comment]
Most of the implementation is done, but I'd like an opinion on the formatting choices. Specifically:
* Changing Arrow to use the curried style
  * I chose this way to disambiguate between n-ary functions and 1-ary tuple functions
  * The effect formatting for this is a little weird.

For most cases you can see examples in the test cases.

Still to be done:
* Type comparison formatter
* Removing `Show` instance (a couple places depend on this, making the change nontrivial)

File: main/test/ca/uwaterloo/flix/language/debug/TestFormatType.scala [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+import ca.uwaterloo.flix.TestUtils
+import ca.uwaterloo.flix.language.ast.{Kind, Rigidity, Symbol, Type, TypeConstructor}
+import org.scalatest.FunSuite
+
+class TestFormatType extends FunSuite with TestUtils {
+  test("FormatType.WellFormedType.Record.External.01") {
+    val tpe = Type.mkRecordExtend("x", Type.Int32, Type.mkRecordExtend("y", Type.Str, Type.RecordEmpty))
+
+    val expected = "{ x: Int32, y: Str }"
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/debug/TestFormatType.scala:11]: Str -> String.

We have deprecated `Str` but not yet removed it.

File: main/test/ca/uwaterloo/flix/language/debug/TestFormatType.scala [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+  test("FormatWellFormedType.Arrow.External.01") {
+    val paramType = Type.Var(0, Kind.Star, Rigidity.Rigid)
+    paramType.setText("t1")
+    val tpe = Type.mkArrow(paramType, Type.Pure, paramType)
+
+    val expected = "t1 -> t1"
+    val actual = FormatType.format(tpe)(Audience.External)
+
+    assert(actual == expected)
+  }
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/debug/TestFormatType.scala:37]: Missing a case where the arrow is impure and should be printed with `~>`.

**Comment by mlutze** [comment]
> I did not understand your comment about curried arrows.

Sorry, should have included an example:

1. An impure arrow of three arguments ends up printing as `Int -> Int -> String ~> String`
2. This previously would have been `(Int, Int, String) ~> String`

I chose this way since with strategy 2 there is no difference between the printed types of
```
def (a: Int, b: Int, c: String): String = ...
```
and
```
def (t: (Int, Int, String)): String = ...
```

However, strategy 1 may obscure the "impure" part a little since the first two arrows make the function appear pure.

**Comment by magnus-madsen** [comment]
OK; that makes sense to me. Its important that we can distinguish those two different types.

**Comment by mlutze** [comment]
Updated!

File: main/src/ca/uwaterloo/flix/language/debug/TypeDiff.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    * Represents a type application.
+    */
+  case class Apply(tpe1: TypeDiff, tpe2: TypeDiff) extends TypeDiff
+
+  /**
+    * Represents two mismatched types.
+    */
+  case class Mismatch(tpe1: Type, tpe2: Type) extends TypeDiff
+
+  def diff(tpe1: Type, tpe2: Type): TypeDiff = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/debug/TypeDiff.scala:44]: Add documentation to outline what the idea is.

File: main/src/ca/uwaterloo/flix/language/debug/FormatType.scala [hunk 4]
```diff
@@ -1,32 +1,309 @@
...
+      }
+      first = false
+    }
+    vt << after
+  }
+
+
+  /**
+    * A flat representation of a schema or record.
+    */
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/debug/FormatType.scala:261]: Add some documentation that represents what this is.

File: main/src/ca/uwaterloo/flix/language/debug/FormatType.scala [hunk 5]
```diff
@@ -1,32 +1,309 @@
...
+            if (applyParams.isEmpty) {
+              applyPart
+            } else {
+              formatApply(s"($applyPart)", applyParams)
+            }
+          }
+
+        case Type.Lambda(tvar, tpe) => audience match {
+          case Audience.Internal => s"${tvar.id.toString} => ${visit(tpe)}"
+          case Audience.External => throw InternalCompilerException(s"Unexpected type: Lambda") // Lambda should not be shown externally
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/debug/FormatType.scala:192]: Hmm, yeah, but maybe crashing is too dangerous.

**Comment by magnus-madsen** [comment]
Not to stress, but it would be nice to merge this soon. Even as-if, since I will soon need to make some changes that would cause more conflicts.

File: main/src/ca/uwaterloo/flix/language/debug/TypeDiff.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  private def mkApply(base: TypeDiff, params: List[TypeDiff]): TypeDiff = {
+    params.foldLeft(base)((base0, param) => TypeDiff.Apply(base0, param))
+  }
+
+  /**
+    * Represents a type constructor.
+    */
+  sealed trait TyCon
```
[the remark this answers is not in the document]
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/debug/TypeDiff.scala:72]: No good reason as far as I can tell. Simplified.

**Comment by magnus-madsen** [comment]
Merged! Awesome :)

**Pull request merged**